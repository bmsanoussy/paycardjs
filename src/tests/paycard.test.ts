import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import Paycard, { CreatePaymentRequest, PaymentMethod } from '../paycard';


describe('Paycard', () => {
    let mock: MockAdapter;
    let paycard: Paycard;

    beforeAll(() => {
        paycard = new Paycard('test-api-key');
    });

    beforeEach(() => {
        mock = new MockAdapter(paycard['client']);
    });

    afterEach(() => {
        mock.reset();
    });

    describe('createPayment', () => {
        it('should throw an error if amount is less than or equal to 0', async () => {
            const request: CreatePaymentRequest = {
                amount: 0,
                description: 'Test payment',
                reference: 'test-ref',
                paymentMethod: PaymentMethod.PAYCARD,
                callbackUrl: 'http://example.com/callback',
                autoRedirect: true,
                redirectWithGet: false
            };

            await expect(paycard.createPayment(request)).rejects.toThrowErrorMatchingSnapshot();
        });

        it('should throw an error if callbackUrl is invalid', async () => {
            const request: CreatePaymentRequest = {
                amount: 100,
                description: 'Test payment',
                reference: 'test-ref',
                paymentMethod: PaymentMethod.PAYCARD,
                callbackUrl: 'invalid-url',
                autoRedirect: true,
                redirectWithGet: false
            };

            await expect(paycard.createPayment(request)).rejects.toThrowErrorMatchingSnapshot();
        });

        it('should create a payment successfully', async () => {
            const request: CreatePaymentRequest = {
                amount: 100,
                description: 'Test payment',
                reference: 'test-ref',
                paymentMethod: PaymentMethod.PAYCARD,
                callbackUrl: 'https://example.com/callback',
                autoRedirect: true,
                redirectWithGet: false
            };

            const response = {
                code: 0,
                payment_amount: 100,
                payment_description: 'Test payment',
                operation_reference: 'test-ref',
                payment_url: 'http://example.com/payment'
            };

            mock
                .onPost('/epay/create')
                .replyOnce(200, response);

            const result = await paycard.createPayment(request);

            expect(mock.history).toMatchSnapshot();
            expect(result).toMatchSnapshot();
        });

        it('should throw an error if paycard response code is not 0', async () => {
            const request: CreatePaymentRequest = {
                amount: 100,
                description: 'Test payment',
                reference: 'test-ref',
                paymentMethod: PaymentMethod.PAYCARD,
                callbackUrl: 'http://example.com/callback',
                autoRedirect: true,
                redirectWithGet: false
            };

            const response = {
                code: -1,
                error_message: 'Invalid amount',
            };

            mock
                .onPost('/epay/create')
                .replyOnce(200, response);

            await expect(paycard.createPayment(request)).rejects.toThrowErrorMatchingSnapshot();
        });

        it('should throw an error if API call fails', async () => {
            const request: CreatePaymentRequest = {
                amount: 100,
                description: 'Test payment',
                reference: 'test-ref',
                paymentMethod: PaymentMethod.PAYCARD,
                callbackUrl: 'http://example.com/callback',
                autoRedirect: true,
                redirectWithGet: false
            };

            mock
                .onPost('/epay/create')
                .networkError();

            await expect(paycard.createPayment(request)).rejects.toThrowErrorMatchingSnapshot();
        });
    });

    describe('getPaymentStatus', () => {
        it('should get payment status successfully', async () => {
            const reference = 'test-ref';
            const response = {
                code: 0,
                status: 'SUCCESS',
                status_description: 'Payment successful',
                error_message: '',
                payment_amount: '100',
                payment_amount_formatted: '100',
                ecommerce_description: 'Test payment',
                reference: 'test-ref'
            };

            mock
                .onGet(`/epay/test-api-key/${reference}/status`)
                .replyOnce(200, response);

            const result = await paycard.getPaymentStatus(reference);

            expect(mock.history).toMatchSnapshot();
            expect(result).toMatchSnapshot();
        });

        it('should throw an error if paycard response code is not 0', async () => {
            const reference = 'test-ref';

            const response = {
                code: -1,
                error_message: 'Unknown reference',
            };

            mock
                .onGet(`/epay/test-api-key/${reference}/status`)
                .replyOnce(200, response);

            await expect(paycard.getPaymentStatus(reference)).rejects.toThrowErrorMatchingSnapshot();
        });

        it('should throw an error if API call fails', async () => {
            const reference = 'test-ref';

            mock
                .onGet(`/epay/test-api-key/${reference}/status`)
                .networkError();

            await expect(paycard.getPaymentStatus(reference)).rejects.toThrowErrorMatchingSnapshot();
        });
    });
});
