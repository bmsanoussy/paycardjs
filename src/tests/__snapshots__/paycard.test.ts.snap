// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paycard createPayment should create a payment successfully 1`] = `
{
  "delete": [],
  "get": [],
  "head": [],
  "link": [],
  "list": [],
  "options": [],
  "patch": [],
  "post": [
    {
      "baseURL": "https://mapaycard.com",
      "data": "{"c":"test-api-key","paycard-amount":100,"paycard-description":"Test payment","paycard-operation-reference":"test-ref","paycard-callback-url":"https://example.com/callback","paycard-auto-redirect":"on","paycard-redirect-with-get":"off","paycard-jump-to-paycard":"on"}",
      "env": {
        "Blob": [Function],
        "FormData": [Function],
      },
      "headers": {
        "Accept": "application/json, text/plain, */*",
        "Content-Type": "application/json",
      },
      "maxBodyLength": -1,
      "maxContentLength": -1,
      "method": "post",
      "timeout": 0,
      "transformRequest": [
        [Function],
      ],
      "transformResponse": [
        [Function],
      ],
      "transitional": {
        "clarifyTimeoutError": false,
        "forcedJSONParsing": true,
        "silentJSONParsing": true,
      },
      "url": "/epay/create",
      "validateStatus": [Function],
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
    },
  ],
  "put": [],
  "unlink": [],
}
`;

exports[`Paycard createPayment should create a payment successfully 2`] = `
{
  "code": 0,
  "operation_reference": "test-ref",
  "payment_amount": 100,
  "payment_description": "Test payment",
  "payment_url": "http://example.com/payment",
}
`;

exports[`Paycard createPayment should throw an error if API call fails 1`] = `"API call failed: Network Error"`;

exports[`Paycard createPayment should throw an error if amount is less than or equal to 0 1`] = `"Amount must be greater than 0"`;

exports[`Paycard createPayment should throw an error if callbackUrl is invalid 1`] = `"Callback URL must be a valid URL"`;

exports[`Paycard createPayment should throw an error if paycard response code is not 0 1`] = `"Invalid amount"`;

exports[`Paycard getPaymentStatus should get payment status successfully 1`] = `
{
  "delete": [],
  "get": [
    {
      "baseURL": "https://mapaycard.com",
      "data": undefined,
      "env": {
        "Blob": [Function],
        "FormData": [Function],
      },
      "headers": {
        "Accept": "application/json, text/plain, */*",
        "Content-Type": "application/json",
      },
      "maxBodyLength": -1,
      "maxContentLength": -1,
      "method": "get",
      "timeout": 0,
      "transformRequest": [
        [Function],
      ],
      "transformResponse": [
        [Function],
      ],
      "transitional": {
        "clarifyTimeoutError": false,
        "forcedJSONParsing": true,
        "silentJSONParsing": true,
      },
      "url": "/epay/test-api-key/test-ref/status",
      "validateStatus": [Function],
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
    },
  ],
  "head": [],
  "link": [],
  "list": [],
  "options": [],
  "patch": [],
  "post": [],
  "put": [],
  "unlink": [],
}
`;

exports[`Paycard getPaymentStatus should get payment status successfully 2`] = `
{
  "code": 0,
  "ecommerce_description": "Test payment",
  "error_message": "",
  "payment_amount": "100",
  "payment_amount_formatted": "100",
  "reference": "test-ref",
  "status": "SUCCESS",
  "status_description": "Payment successful",
}
`;

exports[`Paycard getPaymentStatus should throw an error if API call fails 1`] = `"API call failed: Network Error"`;

exports[`Paycard getPaymentStatus should throw an error if paycard response code is not 0 1`] = `"API call failed: Unknown error"`;
